ProgettoCappelloCosta

LeonardoCapp opened the pull request:
Abbiamo  eseguito l'upload del file rrd_mts_similarity.c per la similarità di serie multivariate e modificato il makefile di conseguenza.

LeonardoCapp committed:
Add files via upload

LeonardoCapp committed:
Update Makefile.in

IvanNardi commented:
@LeonardoCapp, could you:
* rebase on current dev branch? (Makefile has been just updated; the goal is to compile your code in the CI tests)
* fix Codiga warnings (https://github.com/ntop/nDPI/pull/1404/checks?check_run_id=4808452494)? If you want, you can ignore the " Codiga found complex code " and " Codiga found too long functions" stuff.

Thanks

LeonardoCapp committed:
Update rrd_mts_similarity.c

minor warning fixed.

lucaderi reviewed (changes requested):
This code is 99% similar to rrd_similarity. so you must provide a pull request that improves such tool instead of coding a brand new one.

lucaderi commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -0,0 +1,483 @@
+/* *************************************************** */
+
+static void help() {
+  printf("Usage: rrd_similarity [-v][-a <alpha>][-e <end>][-q][-s <start>]\n"
```
Change the name

lucaderi commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -0,0 +1,483 @@
+/*
+ * rrd_similarity.c
```
Change the name

lucaderi commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -0,0 +1,483 @@
+	 ,
+	 DEFAULT_ALPHA, DEFAULT_END, DEFAULT_START, similarity_threshold);
+
+  printf("\n\nExample: rrd_similarity -q -f bytes.rrd -d /var/lib/ntopng/-1/snmpstats\n");
```
Change the name

lucaderi commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -0,0 +1,483 @@
+        analyze_rrd(&rrdms[i].rfs[j], start, end);
+    }
+
+    analyze_mts(&rrdms[i], start, end, n_file);
```
How to do you specify the RRDs to analzyze? A new command line option is requested as well you need to improve the help

lucaderi reviewed (changes requested):
This pull request cannot be merged until you have made all the requested fixed.

LeonardoCapp committed:
Delete rrd_similarity.c

LeonardoCapp committed:
Add files via upload

LeonardoCapp committed:
Update Makefile.in

LeonardoCapp committed:
Merge pull request #1 from LeonardoCapp/LeonardoCapp-patch-1

Delete rrd_similarity.c

LeonardoCapp commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -58,22 +59,22 @@ u_int verbose = 0, similarity_threshold = 100, skip_zero = 0;
 static void help() {
-  printf("Usage: rrd_similarity [-v][-a <alpha>][-e <end>][-q][-s <start>]\n"
-	 "                      -f <filename> -d <basedir> [-t <threshold>]\n"
+  printf("Usage: rrd_mts_similarity [-v][-a <alpha>][-e <end>][-q][-s <start>]\n"
```
updated example, rrd_similarity updated to rrd_mts_similarity

lucaderi replied:
What is the difference here?

LeonardoCapp commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -58,22 +59,22 @@ u_int verbose = 0, similarity_threshold = 100, skip_zero = 0;
 
 	 "-d <basedir>   | Base directory where RRD filename is searched\n"
-	 "-f <rrd path>  | Path of the RRD filename to analyze\n"
+	 "-f <rrd_path1> <rrd_path2> ...  | Path of the RRDs filename to analyze\n"
```
new explanation of how -f works

lucaderi replied:
Formatting is broken and <rrd_path1> <rrd_path2> should be [<rrd_path>]+

LeonardoCapp replied:
Reworked -f in commit 108daa5

LeonardoCapp commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -58,22 +59,22 @@ u_int verbose = 0, similarity_threshold = 100, skip_zero = 0;
 	 DEFAULT_ALPHA, DEFAULT_END, DEFAULT_START, similarity_threshold);
 
-  printf("\n\nExample: rrd_similarity -q -f bytes.rrd -d /var/lib/ntopng/-1/snmpstats\n");
+  printf("\n\nExample: rrd_mts_similarity -q -f bytes.rrd score.rrd -d /var/lib/ntopng/-1/snmpstats\n");
```
updated example 
from rrd_similarity to rrd_mts_similarity
from -f bytes.rrd to -> -f bytes.rrd score.rrd

lucaderi replied:
Without " "  (quotes) it cannot work

LeonardoCapp replied:
Reworked -f in commit 108daa5

LeonardoCapp commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -363,63 +364,57 @@ int main(int argc, char *argv[]) {
   }
 
-  if((filename == NULL) || (basedir == NULL))
+  if((filename[0] == NULL) || (basedir == NULL))
```
if filename[0] is null no file has been passed on cli

LeonardoCapp commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -363,63 +364,57 @@ int main(int argc, char *argv[]) {
+    filename[i] = NULL;
+  }
+  
+  for(i= 0; i<argc;i++)
```
Used for the parsing of the arguments instead of while((c = getopt(argc, argv, "d:s:e:a:qf:t:vz")) != '?'),  improved -f

lucaderi replied:
Not a good idea. Please restore it.

LeonardoCapp replied:
Fixed in commit with ID 030a0c3

LeonardoCapp committed:
Update rrd_mts_similarity.c

89 removed unused variable j 
126 removed unused variable p

LeonardoCapp committed:
Update rrd_mts_similarity.c

Improved output if file not found in find_rrd_similarity

LeonardoCapp committed:
Update rrd_mts_similarity.c

removed unused parameter from function find_rrd_multi_similarities

LeonardoCapp committed:
Update rrd_mts_similarity.c

Corrected analyze_mts bin computing, now an accumulator stores the value and then is inserted in the bin
removed if
 introduced variable not_found, counts any file has no correspondences in the subdir examined

lucaderi commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -0,0 +1,483 @@
+  for(i=0; i<MAX_NUM_FILE; i++)
+  {
+    filename[i] = malloc(sizeof(char)*50);
+    filename[i] = NULL;
```
This code has no sense (first allocate then set it to zero). Consider using strdup()

LeonardoCapp replied:
Fixed in commit with ID 030a0c3

lucaderi commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -0,0 +1,483 @@
+  if(t)
+    return 1;
+  else
+    return 0;
```
The return code is never checked by the caller. Furthermore you must find all rrds per host otherwise the code wont work

LeonardoCapp replied:
Finding all rrds is done in commit 80a751a.
Return type is modified in 2d1460c.

lucaderi commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -0,0 +1,483 @@
+	  	) {
+	  float similarity = ndpi_bin_similarity(&rrdms[i].rfs[k].b, &rrdms[j].rfs[k].b, 0, similarity_threshold);
+
+	  if((similarity >= 0) && (similarity < similarity_threshold)) {
```
The cumulative similarity is not considered, but only individual. Not what I expect

LeonardoCapp replied:
Fixed in commit with ID bcdca2f

lucaderi commented on rrdtool/rrd_mts_similarity.c:
```diff
@@ -0,0 +1,483 @@
+
+  ndpi_free(rrdms);
+
+  return(0);
```
filename memory is never freed

LeonardoCapp replied:
Fixed in commit with ID 030a0c3

lucaderi reviewed (changes requested):
Many things are still pending before this code can be merged

LeonardoCapp committed:
Update rrd_mts_similarity.c

Restored while for parsing arguments
Fixed missing free filename pointer
Fixed the initializzation of filename[i] using strdup instead of malloc
At line 430 the guard was modified from filename[0] == NULL to filename == NULL

LeonardoCapp committed:
Update rrd_mts_similarity.c

reworked -f, now the arguments are separated using '+' character. Strtok is used for parsing them.
Help method has been updated to match the new -f changes.

LeonardoCapp committed:
Update rrd_mts_similarity.c

Now we consider an host only if his directory contains all the requested files.

LeonardoCapp committed:
Update rrd_mts_similarity.c

Removed reduntant vector num_rrds.
Removed num_rrds argument from methods find_rrds and find_rrd_similarities invocation.

LeonardoCapp committed:
Update rrd_mts_similarity.c

Removed unused variable not_found from analyze_mts

LeonardoCapp committed:
Update rrd_mts_similarity.c

Merged the methods find_rrd_similarities with find_rrd_multi_similarities.
find_rrd_similarities now check the similarity of two file at a time, and it's called when two multivariate time series are similar.

lucaderi commented:
@LeonardoCapp When you update the pull request you need to reply to all the individual comments without repushing everything in one shot as I need to see what you have changed

LeonardoCapp committed:
Update rrd_mts_similarity.c

Changed find_rrd's return type from int to void.

LeonardoCapp committed:
Update rrd_mts_similarity.c

Replaced "+" with "," in method help and in f case in main function.

LeonardoCapp committed:
Update rrd_mts_similarity.c

Line 52 added new attribute path.
Line 236 and 247, the output now returns host's path.
Line 272, modified *rrdms in **rrdms.
Line 319, initialization of host path.
Line 434, used &rrdms instead of rrdms.

lucaderi commented:
This PR is idle for a long time and not mergeable with the current code. Please update it and resubmit it in a clean state.
